src/main/java/com/fauna/codec/CodecRegistryKey.java

```diff
@@ -1,16 +1,38 @@
 package com.fauna.codec;
 
+import java.lang.reflect.Type;
+import java.util.Objects;
+
 public class CodecRegistryKey {
     private final Class<?> base;
-    private final Class<?> subtype;
-    public <T> CodecRegistryKey(Class<?> clazz, Class<?> subtype) {
+    private final Type typeArg;
+    public <T> CodecRegistryKey(Class<T> clazz, Type typeArg) {
         base = clazz;
-        this.subtype = subtype;
+        this.typeArg = typeArg;
+    }
+
+    public static <T> CodecRegistryKey from(Class<T> clazz) {
+        return new CodecRegistryKey(clazz, null);
     }
 
-    public static <T> CodecRegistryKey from(Class<?> clazz, Class<?> subtype) {
-        return new CodecRegistryKey(clazz, subtype);
+    public static <T> CodecRegistryKey from(Class<T> clazz, Type typeArg) {
+        return new CodecRegistryKey(clazz, typeArg);
     }
 
-    // TODO: Override equals/hash
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        } else if (other instanceof CodecRegistryKey) {
+            CodecRegistryKey otherCRK = (CodecRegistryKey) other;
+            return this.base == otherCRK.base && this.typeArg == otherCRK.typeArg;
```

Comment:
I _think_ you want to call equals here

```suggestion
            return this.base.equals(otherCRK.base) && this.typeArg.equals(otherCRK.typeArg);
```

Reply:
Good catch. I did something slightly different because the typeArg can be null.